<a href="https://colab.research.google.com/github/SARA3SAEED/more-practice/blob/main/DA_Le_Lab11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Shop orders analysis

In this challenge, you will analyze data from [Sumup](https://www.sumup.com/fr-fr/) who wants to analyse its restaurant orders and categorize them.


Upload [Tiller_order_data.csv](https://drive.google.com/file/d/1dVuD1hB3UnX5xhCHFMQ5R4iBdMdJMK4a/view?usp=sharing) in Colab and load it in a DataFrame `df_order`


## Exploratory Analysis


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
#your code here
import pandas as pd
df_order = pd.read_csv('/content/drive/MyDrive/dataset/le/week5-day2-lab3/Tiller_order_data.csv')
df_order.head()

<ipython-input-3-87df936982bc>:3: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  df_order = pd.read_csv('/content/drive/MyDrive/dataset/le/week5-day2-lab3/Tiller_order_data.csv')


id_order  id_store  id_table  id_waiter  id_customer  \
0  55538867      8052       NaN        NaN          NaN   
1  56035309      8052       NaN        NaN          NaN   
2  55550051      8052       NaN        NaN          NaN   
3  57000119      8052       NaN    16199.0          NaN   
4  55558817      8052       NaN        NaN          NaN   

                            id_external  id_device              date_opened  \
0  0425716B-EFF4-41CA-AEA1-839104F36833    15327.0  2019-01-12 13:02:17 UTC   
1  75E41FE2-64FF-41D3-954C-A7DE4AA887EF    15327.0  2019-01-16 19:39:09 UTC   
2  F6051A05-C9AC-4033-BF72-BB5149B8F439    15327.0  2019-01-12 14:18:46 UTC   
3  B8BEEC66-1C10-48A0-B4D5-035CB5EEFE62    15327.0  2019-01-24 17:49:12 UTC   
4  17F0533C-2FF1-4FC5-A50D-12704C7B7A4B    15327.0  2019-01-12 15:25:06 UTC   

               date_closed                    dim_name dim_status  dim_type  \
0  2019-01-12 19:58:38 UTC                     vincent     CLOSED         1   
1  2019-01-16 22:10:50 UTC  frere et soeur avec pierre     CLOSED         1   
2  2019-01-12 19:50:32 UTC                      rachel     CLOSED         1   
3  2019-01-24 21:58:59 UTC                  Groupe PEL     CLOSED         1   
4  2019-01-12 19:21:03 UTC                remi et date     CLOSED         1   

  dim_comment    dim_source  m_nb_customer  m_cached_payed  m_cached_price  
0         NaN   Tiller iPAD              1            45.5            45.5  
1         NaN   Tiller iPAD              2            49.8            49.8  
2         NaN   Tiller iPAD              1            27.4            27.4  
3         NaN  tiller-order              3            60.0            60.0  
4         NaN   Tiller iPAD              2            39.9            39.9

What does a row represent? What information can you find?

> **Your Answer Here (Double click)**


What are the different statuses of an order? How many orders are there per status?   

In [5]:
#your code here
df_order.dim_status.value_counts()

dim_status
CLOSED         116939
IN_PROGRESS       365
Name: count, dtype: int64

What is the proportion of orders ?  \
You can use the `.value_counts()` function, with the optional argument `normalize=True`

In [6]:
#your code here
df_order.dim_status.value_counts(normalize=True)

dim_status
CLOSED         0.996888
IN_PROGRESS    0.003112
Name: proportion, dtype: float64

How many restaurants are there? What are their ids?

In [7]:
#your code here
df_order.id_store.nunique()

10

In [15]:
df_order.id_store.unique()

array([8052, 8347, 8283, 9084, 4337,  360, 1796, 7965, 6293, 4542])

In [9]:
df_order.id_store.value_counts()

id_store
1796    27349
4337    20003
4542    17641
6293    16103
8283    10023
7965     9091
9084     7804
8052     4906
360      4383
8347        1
Name: count, dtype: int64

## Aggregations

Compute the total turnover per restaurant. Which restaurant has the highest turnover?


In [18]:
#your code here
re_turn = df_order.groupby('id_store')['m_cached_price'].sum()
re_turn

id_store
360      386934.14
1796     321099.69
4337     865847.30
4542     243789.69
6293     359718.00
7965    1138585.95
8052     327758.85
8283     369822.82
8347         23.00
9084     479040.75
Name: m_cached_price, dtype: float64

In [25]:
re_turn.max()

1138585.95

In [27]:
re_turn.idxmax()

7965

Count the number of orders per restaurant. Which restaurant has the most orders? Is it the same as the one with the highest turnover?


In [29]:
#your code here
order=df_order.groupby('id_store')['id_order'].count()
order

id_store
360      4383
1796    27349
4337    20003
4542    17641
6293    16103
7965     9091
8052     4906
8283    10023
8347        1
9084     7804
Name: id_order, dtype: int64

In [30]:
order.max()

27349

Calculate the average price of an order for each restaurant. Do you think you are dealing with similar categories of restaurants ?


In [32]:
#your code here
df_order.groupby('id_store')['m_cached_price'].mean()

id_store
360      88.280662
1796     11.740820
4337     43.285872
4542     13.819494
6293     22.338570
7965    125.243202
8052     66.807756
8283     36.897418
8347     23.000000
9084     61.384002
Name: m_cached_price, dtype: float64

## Column transformations


The problem is that an order may be placed for several people. Instead, we want to know the average price spent by a single person in the restaurant.

Add a column to `df_order`,  representing the price per customer for each order.


In [33]:
#your code here
df_order['price_per_customer']=df_order['m_cached_price']/df_order['m_nb_customer']
df_order.head()

id_order  id_store  id_table  id_waiter  id_customer  \
0  55538867      8052       NaN        NaN          NaN   
1  56035309      8052       NaN        NaN          NaN   
2  55550051      8052       NaN        NaN          NaN   
3  57000119      8052       NaN    16199.0          NaN   
4  55558817      8052       NaN        NaN          NaN   

                            id_external  id_device              date_opened  \
0  0425716B-EFF4-41CA-AEA1-839104F36833    15327.0  2019-01-12 13:02:17 UTC   
1  75E41FE2-64FF-41D3-954C-A7DE4AA887EF    15327.0  2019-01-16 19:39:09 UTC   
2  F6051A05-C9AC-4033-BF72-BB5149B8F439    15327.0  2019-01-12 14:18:46 UTC   
3  B8BEEC66-1C10-48A0-B4D5-035CB5EEFE62    15327.0  2019-01-24 17:49:12 UTC   
4  17F0533C-2FF1-4FC5-A50D-12704C7B7A4B    15327.0  2019-01-12 15:25:06 UTC   

               date_closed                    dim_name dim_status  dim_type  \
0  2019-01-12 19:58:38 UTC                     vincent     CLOSED         1   
1  2019-01-16 22:10:50 UTC  frere et soeur avec pierre     CLOSED         1   
2  2019-01-12 19:50:32 UTC                      rachel     CLOSED         1   
3  2019-01-24 21:58:59 UTC                  Groupe PEL     CLOSED         1   
4  2019-01-12 19:21:03 UTC                remi et date     CLOSED         1   

  dim_comment    dim_source  m_nb_customer  m_cached_payed  m_cached_price  \
0         NaN   Tiller iPAD              1            45.5            45.5   
1         NaN   Tiller iPAD              2            49.8            49.8   
2         NaN   Tiller iPAD              1            27.4            27.4   
3         NaN  tiller-order              3            60.0            60.0   
4         NaN   Tiller iPAD              2            39.9            39.9   

   price_per_customer  
0               45.50  
1               24.90  
2               27.40  
3               20.00  
4               19.95

Using the previous column, calculate the average price spent by each person in each of the restaurants.

In [34]:
#your code here
df_order.groupby('id_store')['price_per_customer'].mean()

id_store
360           inf
1796          inf
4337          inf
4542    13.819494
6293          inf
7965          inf
8052    16.698658
8283    36.897418
8347    11.500000
9084          inf
Name: price_per_customer, dtype: float64

There seems to be problem with the quality of data, as we can notice the presence of `NaN`for half of our restaurant

Do you have any idea why?


> **Your Answer Here (Double click)**

## Cleaning


Fill the 0 values in the `m_nb_customer` column with 1, then eliminate the rows where the `m_cached_price` is 0.

In [35]:
#clean m_nb_customer column
#your code here
df_order.loc[df_order['m_nb_customer']==0,'m_nb_customer']=1
df_order.head()

id_order  id_store  id_table  id_waiter  id_customer  \
0  55538867      8052       NaN        NaN          NaN   
1  56035309      8052       NaN        NaN          NaN   
2  55550051      8052       NaN        NaN          NaN   
3  57000119      8052       NaN    16199.0          NaN   
4  55558817      8052       NaN        NaN          NaN   

                            id_external  id_device              date_opened  \
0  0425716B-EFF4-41CA-AEA1-839104F36833    15327.0  2019-01-12 13:02:17 UTC   
1  75E41FE2-64FF-41D3-954C-A7DE4AA887EF    15327.0  2019-01-16 19:39:09 UTC   
2  F6051A05-C9AC-4033-BF72-BB5149B8F439    15327.0  2019-01-12 14:18:46 UTC   
3  B8BEEC66-1C10-48A0-B4D5-035CB5EEFE62    15327.0  2019-01-24 17:49:12 UTC   
4  17F0533C-2FF1-4FC5-A50D-12704C7B7A4B    15327.0  2019-01-12 15:25:06 UTC   

               date_closed                    dim_name dim_status  dim_type  \
0  2019-01-12 19:58:38 UTC                     vincent     CLOSED         1   
1  2019-01-16 22:10:50 UTC  frere et soeur avec pierre     CLOSED         1   
2  2019-01-12 19:50:32 UTC                      rachel     CLOSED         1   
3  2019-01-24 21:58:59 UTC                  Groupe PEL     CLOSED         1   
4  2019-01-12 19:21:03 UTC                remi et date     CLOSED         1   

  dim_comment    dim_source  m_nb_customer  m_cached_payed  m_cached_price  \
0         NaN   Tiller iPAD              1            45.5            45.5   
1         NaN   Tiller iPAD              2            49.8            49.8   
2         NaN   Tiller iPAD              1            27.4            27.4   
3         NaN  tiller-order              3            60.0            60.0   
4         NaN   Tiller iPAD              2            39.9            39.9   

   price_per_customer  
0               45.50  
1               24.90  
2               27.40  
3               20.00  
4               19.95

In [36]:
#clean m_cached_price column
df_order.loc[df_order['m_cached_price']==0]


id_order  id_store  id_table  id_waiter  id_customer  \
206      88608012       360       NaN        NaN    1822855.0   
207     106037340       360       NaN        NaN    2098543.0   
208     104251222       360       NaN        NaN    2065980.0   
209      70973591       360       NaN        NaN    1539225.0   
210      98319978       360       NaN        NaN    1973862.0   
...           ...       ...       ...        ...          ...   
97639    79700030      1796       NaN     2418.0          NaN   
97641    66270090      1796       NaN     2418.0          NaN   
98339    79536673      1796       NaN     2418.0          NaN   
104982   65215932      4542       NaN        NaN          NaN   
113804   65231560      4542       NaN        NaN          NaN   

                                 id_external  id_device  \
206     2ad04d78-a5eb-4224-91f8-9a5f2ee9a051        NaN   
207     2c221e7f-87a8-4995-bc56-fa8a21a1693b        NaN   
208     bdb334c7-0417-471c-b0b6-f7545c5732a9        NaN   
209     a5b23198-2d7c-46dd-ab2d-ec7125d8fd15        NaN   
210     69280e86-de0a-4cf1-bcde-4764258e3489        NaN   
...                                      ...        ...   
97639   459E7F39-6462-47E5-B9D3-000745576927    17383.0   
97641   2CE6294B-F526-4D97-86FE-835A70C8C958    17383.0   
98339   B2C92FAE-F98E-457B-A28F-A5998D60467E    17383.0   
104982  4BE42681-CA68-41D1-B503-9B357062C3BD     7411.0   
113804  2814FA60-8877-40FA-B7A1-863C9A5AEDE3     7411.0   

                    date_opened              date_closed  \
206     2019-08-22 18:16:11 UTC  2019-08-22 18:26:30 UTC   
207     2019-11-30 17:14:22 UTC  2019-11-30 18:58:51 UTC   
208     2019-11-20 12:59:03 UTC  2019-11-20 19:51:22 UTC   
209     2019-05-03 17:37:50 UTC  2019-05-03 19:10:58 UTC   
210     2019-10-16 08:15:34 UTC  2019-10-16 17:55:52 UTC   
...                         ...                      ...   
97639   2019-06-28 10:41:00 UTC  2019-06-28 10:42:47 UTC   
97641   2019-04-02 10:30:09 UTC  2019-04-03 10:27:34 UTC   
98339   2019-06-27 11:36:00 UTC  2019-06-27 11:38:41 UTC   
104982  2019-03-26 11:35:49 UTC  2019-03-26 11:37:00 UTC   
113804  2019-03-26 12:09:45 UTC  2019-03-26 12:11:29 UTC   

                              dim_name   dim_status  dim_type dim_comment  \
206     LaFourchette -20% sur la carte  IN_PROGRESS         1         NaN   
207     LaFourchette -20% sur la carte  IN_PROGRESS         1         NaN   
208     LaFourchette -20% sur la carte  IN_PROGRESS         1         NaN   
209     LaFourchette -20% sur la carte  IN_PROGRESS         1         NaN   
210                       LaFourchette  IN_PROGRESS         1         NaN   
...                                ...          ...       ...         ...   
97639                              L28       CLOSED         1         NaN   
97641                              L20       CLOSED         1         NaN   
98339                              L80       CLOSED         1         NaN   
104982                             NaN       CLOSED         1         NaN   
113804                             NaN       CLOSED         1         NaN   

          dim_source  m_nb_customer  m_cached_payed  m_cached_price  \
206     LAFOURCHETTE              2             0.0             0.0   
207     LAFOURCHETTE              2             0.0             0.0   
208     LAFOURCHETTE              2             0.0             0.0   
209     LAFOURCHETTE              3             0.0             0.0   
210     LAFOURCHETTE              3             0.0             0.0   
...              ...            ...             ...             ...   
97639   tiller-order              1             0.0             0.0   
97641   tiller-order              1             0.0             0.0   
98339   tiller-order              1             0.0             0.0   
104982   Tiller iPAD              1             0.0             0.0   
113804   Tiller iPAD              1             0.0             0.0   

        price_per_cu

Recalculate the price_per_customer column and the average price spent by each person in each restaurant.

In [56]:
#your code here
df_order['price_per_customer']=df_order['m_cached_price']/df_order['m_nb_customer']
df_order.head(2)

id_order  id_store  id_table  id_waiter  id_customer  \
0  55538867      8052       NaN        NaN          NaN   
1  56035309      8052       NaN        NaN          NaN   

                            id_external  id_device               date_opened  \
0  0425716B-EFF4-41CA-AEA1-839104F36833    15327.0 2019-01-12 13:02:17+00:00   
1  75E41FE2-64FF-41D3-954C-A7DE4AA887EF    15327.0 2019-01-16 19:39:09+00:00   

                date_closed                    dim_name dim_status  dim_type  \
0 2019-01-12 19:58:38+00:00                     vincent     CLOSED         1   
1 2019-01-16 22:10:50+00:00  frere et soeur avec pierre     CLOSED         1   

  dim_comment   dim_source  m_nb_customer  m_cached_payed  m_cached_price  \
0         NaN  Tiller iPAD              1            45.5            45.5   
1         NaN  Tiller iPAD              2            49.8            49.8   

   price_per_customer  order_duration  
0                45.5      416.350000  
1                24.9      151.683333

In [38]:
df_order.groupby('id_store')['price_per_customer'].mean()

id_store
360      29.022309
1796     11.740820
4337     43.282793
4542     13.819494
6293     16.805911
7965    102.411383
8052     16.698658
8283     36.897418
8347     11.500000
9084     30.452830
Name: price_per_customer, dtype: float64

## Order duration analysis


Restaurant `7965` has a higher price per person than the rest.  
Let's look at the time spent at the table to determine if it matches the characteristics of a fancy restaurant.




Convert the `date_opened` and `date_closed` columns to **datetime** format.


In [39]:
#your code here
df_order['date_opened']=pd.to_datetime(df_order['date_opened'])
df_order['date_closed']=pd.to_datetime(df_order['date_closed'])
df_order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 117304 entries, 0 to 117303
Data columns (total 18 columns):
 #   Column              Non-Null Count   Dtype              
---  ------              --------------   -----              
 0   id_order            117304 non-null  int64              
 1   id_store            117304 non-null  int64              
 2   id_table            21917 non-null   float64            
 3   id_waiter           57377 non-null   float64            
 4   id_customer         1867 non-null    float64            
 5   id_external         117232 non-null  object             
 6   id_device           108906 non-null  float64            
 7   date_opened         117304 non-null  datetime64[ns, UTC]
 8   date_closed         117304 non-null  datetime64[ns, UTC]
 9   dim_name            56553 non-null   object             
 10  dim_status          117304 non-null  object             
 11  dim_type            117304 non-null  int64              
 12  dim_comment     

Run the code below to calculate the time between the opening and closing date of an order in minutes.

In [40]:
df_order["order_duration"] = (df_order.date_closed - df_order.date_opened).dt.total_seconds() / 60

Compute the average duration spent at each table for every restaurant.

In [41]:
#your code here
df_order.groupby('id_store')['order_duration'].mean()

id_store
360     274.415495
1796      3.437423
4337      6.240056
4542      3.930243
6293     30.246884
7965     53.703337
8052    164.686744
8283      2.005060
8347      3.900000
9084    125.554261
Name: order_duration, dtype: float64

Certain restaurants exhibit notably brief average order times, indicating a potential presence of fast food establishments among them.


## Joins

Download  [Tiller_order_line.csv](https://drive.google.com/file/d/1KSDFGLu2QAAR2C5SUlM5mfhINHU-LIuJ/view?usp=sharing) and [Tiller_store_data.csv](https://drive.google.com/file/d/1W032PaySL7gouyX3bEKgndtm6qVta7Qp/view?usp=sharing) and upload them in two DataFrame `df_line` and `df_store`.


In [42]:
#your code here
df_line = pd.read_csv('/content/drive/MyDrive/dataset/le/week5-day2-lab3/Tiller_order_line.csv')
df_store = pd.read_csv('/content/drive/MyDrive/dataset/le/week5-day2-lab3/Tiller_store_data.csv')

What new information do you have access to through this data?

In [44]:
#your code here
df_line.head(2)

id_order_line   id_order             date_opended             date_created  \
0      626609716  139058632  2020-08-25 17:27:14 UTC  2020-08-25 17:27:25 UTC   
1      403167240   94770355  2019-09-27 07:47:14 UTC  2019-09-27 07:47:18 UTC   

   m_quantity  m_unit_price  m_unit_price_exc_vat  m_total_price_inc_vat  \
0         1.0          15.0             12.500000                   15.0   
1         1.0          25.9             21.583333                   25.9   

   m_total_price_exc_vat  m_tax_percent  m_discount_amount dim_type  \
0              12.500000            0.2                0.0  PRODUCT   
1              21.583333            0.2                0.0  PRODUCT   

           dim_category                              dim_name dim_status  \
0           VINS ROUGES  TWIST AGAIN 2018 VDF - YOANN ROMAGNE       DONE   
1  LIQUIDES 50ML AUTRES                            MOMMY CAKE       DONE   

   m_nb_payed dim_feature_type dim_unit_measure dim_unit_measure_display  
0           0              NaN             UNIT                     UNIT  
1           0              NaN             UNIT                     UNIT

In [45]:
df_store.head(2)

id_store               date_created dim_zipcode dim_country dim_currency
0      2576  2016-07-12 16:15:31+00:00       34120          FR          EUR
1      2538  2016-07-07 16:35:44+00:00       47800          FR          EUR

<details>
    <summary> Answer 💡 </summary>

- `Order line` = The different products in an order
- `Store data` = Information about the restaurant/bar

</details>


Paris is divided into 20 districts. Each of them has a unique code 75001, 75002, ..., 75019, 75020. \
How many orders were placed in the `2nd district`?



In [58]:
# Step 1: Filter stores in the 2nd district
stores_in_2nd_district = df_store[df_store['dim_zipcode'] == '75002']['id_store']

In [63]:
# Step 2: Filter orders based on these stores
orders_in_2nd_district = df_order[df_order['id_store'].isin(stores_in_2nd_district)]

In [62]:
# Step 3: Count the number of orders
num_orders_in_2nd_district = orders_in_2nd_district.id_order.count()
num_orders_in_2nd_district

59446

Let's have a look at restaurant `4542`, which shows a very brief average ordering time.


Retrieve the most ordered items in this restaurant by retaining only the **id_store** and **id_order** columns from `df_order`, as well as the **id_order**, **id_order_line**, and **dim_name** columns from `order_line`.


In [65]:
# Step 1: Filter df_order to include only orders from restaurant 4542
orders_from_4542 = df_order[df_order['id_store'] == 4542][['id_store', 'id_order']]



In [66]:
# Step 2: Merge df_order with df_line on 'id_order' to get the order details
merged_data = orders_from_4542.merge(df_line[['id_order', 'id_order_line', 'dim_name']], on='id_order')



In [67]:
# Step 3: Group by 'dim_name' to count the number of times each item was ordered
most_ordered_items = merged_data.groupby('dim_name').size().reset_index(name='order_count')



In [68]:
# Step 4: Sort the items by order count in descending order
most_ordered_items_sorted = most_ordered_items.sort_values(by='order_count', ascending=False)



In [69]:
# Display the result
most_ordered_items_sorted.head()

dim_name  order_count
51          Plat + Dessert + Boisson         7546
70            Tapas + Plat + Boisson         4745
71  Tapas + Plat + Dessert + Boisson         2978
50                              Plat          795
75                     Tortilla PdeT          481